adapter/oci/app_client.go

```diff
@@ -0,0 +1,208 @@
+}
+
+func (a AppClient) CreateApp(app *adapter.App) (*adapter.App, error) {
+	compartmentId := viper.GetString("oracle.compartment-id")
+
+	subnetIds, err := parseSubnetIds(app.Annotations)
+
+	if err != nil {
+		return nil, err
+	}
+	body := functions.CreateApplicationDetails{
+		CompartmentId: &compartmentId,
+		Config:        app.Config,
+		DisplayName:   &app.Name,
+		SubnetIds:     subnetIds,
+	}
+	req := functions.CreateApplicationRequest{CreateApplicationDetails: body,}
+
+	res, err := a.client.CreateApplication(context.Background(), req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	adapterApp := convertOCIAppToAdapterApp(&res.Application)
+	return adapterApp, nil
+}
+
+func parseSubnetIds(annotations map[string]interface{}) ([]string, error) {
+	if annotations == nil || len(annotations) == 0 {
+		return nil, errors.New("Missing subnets annotation")
+	}
+
+	var subnets []string
+	subnetsInterface, ok := annotations[AnnotationSubnet]
+	if !ok {
+		return nil, errors.New("Missing subnets annotation")
+	}
+
+	// Typecast to []interface{}
+	subnetsArray, success := subnetsInterface.([]interface{})
+
+	if !success {
+		return nil, errors.New("Invalid subnets annotation")
+	}
+
+	for _,s := range subnetsArray {
+		// Typecast to string
+		subnetString, secondSuccess := s.(string)
+		if !secondSuccess {
+			return nil, errors.New("Invalid subnets annotation")
+		}
+		subnets = append(subnets, subnetString)
+	}
+
+	return subnets, nil
+}
+
+func (a AppClient) GetApp(appName string) (*adapter.App, error) {
+	compartmentId := viper.GetString("oracle.compartment-id")
+	req := functions.ListApplicationsRequest{CompartmentId: &compartmentId, DisplayName: &appName}
+	resp, err := a.client.ListApplications(context.Background(), req)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Items) > 0 {
+		getreq := functions.GetApplicationRequest{ApplicationId: resp.Items[0].Id}
+		getres,geterr := a.client.GetApplication(context.Background(), getreq)
+
+		if geterr != nil {
+			return nil, geterr
+		}
+
+		adapterApp := convertOCIAppToAdapterApp(&getres.Application)
+		return adapterApp, nil
+	} else {
+		return nil, adapter.AppNameNotFoundError{Name: appName}
+	}
+}
+
+func (a AppClient) UpdateApp(app *adapter.App) (*adapter.App, error) {
+	//merge config
+	for k, v := range app.Config {
+		if v == "" {
+			delete(app.Config, k)
+		} else {
+			app.Config[k] = v
+		}
+	}
+
+	body := functions.UpdateApplicationDetails{
+		Config: app.Config,
+	}
+
+	req := functions.UpdateApplicationRequest{UpdateApplicationDetails: body, ApplicationId: &app.ID}
+	res, err := a.client.UpdateApplication(context.Background(), req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	adapterApp := convertOCIAppToAdapterApp(&res.Application)
+	return adapterApp, nil
+}
+
+func (a AppClient) DeleteApp(appID string) error {
+	req := functions.DeleteApplicationRequest{ApplicationId: &appID,}
+	_, err := a.client.DeleteApplication(context.Background(), req)
```

Comment:
Should we be checking responses here (and throughout)?

Reply:
Based on the Fn examples in the OCI GO SDK, this is the recommended way to do it.
https://github.com/oracle/oci-go-sdk/blob/master/example/example_functions_test.go#L146